File: jFlash/Classes/CardPeer.m
#import "CardPeer.h"

@implementation CardPeer

//--------------------------------------------------------------------------
// NSMutableArray searchCardsForKeyword
// Returns an array of Cards for a given search string
//--------------------------------------------------------------------------
+ (NSMutableArray*) searchCardsForKeyword: (NSString*) keyword doSlowSearch:(BOOL)slowSearch
{
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  NSMutableArray *cardList = [[[NSMutableArray alloc] init] autorelease];
  NSString* sql;
  FMResultSet *rs;
  NSString* keywordWildcard;
  
  if (!slowSearch){
    int queryLimit = 100;
    int queryLimit2;
    keywordWildcard = [keyword stringByReplacingOccurrencesOfString:@"?" withString:@"*"];

    // Do the search using SQLite FTS (PTAG results)
    sql = [NSString stringWithFormat:@""
           "SELECT c.*, ch.meaning, 0 as card_level, 0 as user_id FROM cards c, cards_html ch "
           "WHERE c.card_id = ch.card_id AND c.card_id in (SELECT card_id FROM cards_search_content WHERE content MATCH '%@' AND ptag = 1 LIMIT %d) "
           "ORDER BY c.headword", keywordWildcard, queryLimit];
    rs = [[db dao] executeQuery:sql];
    int cardListCount = 0;
    while ([rs next]) {
      cardListCount++;
      Card* tmpCard = [[[Card alloc] init] autorelease];
      [tmpCard hydrate:rs];
      [cardList addObject: tmpCard];
    }

    if(cardListCount < queryLimit)
      queryLimit2 = (queryLimit-cardListCount)+queryLimit;
    else
      queryLimit2 = queryLimit;
      
    // Do the search using SQLite FTS (NON-PTAG results)
    sql = [NSString stringWithFormat:@""
           "SELECT c.*, ch.meaning, 0 as card_level, 0 as user_id FROM cards c, cards_html ch "
           "WHERE c.card_id = ch.card_id AND c.card_id in (SELECT card_id FROM cards_search_content WHERE content MATCH '%@' AND ptag = 0 LIMIT %d) "
           "ORDER BY c.headword", keywordWildcard, queryLimit2];
    rs = [[db dao] executeQuery:sql];
    while ([rs next]) {
      Card* tmpCard = [[[Card alloc] init] autorelease];
      [tmpCard hydrate:rs];
      [cardList addObject: tmpCard];
    }
    [rs close];
  }
  else
  {
    // Do slow substring match (w/ ASTERISK)
    keywordWildcard = [keyword stringByReplacingOccurrencesOfString:@" " withString:@"* "];
    sql = [NSString stringWithFormat:@""
           "SELECT c.*, ch.meaning, 0 as card_level, 0 as user_id FROM cards c, cards_html ch "
           "WHERE c.card_id = ch.card_id AND c.card_id in (SELECT card_id FROM cards_search_content WHERE content MATCH '%@*' AND ptag = 0 LIMIT %d) "
           "ORDER BY c.headword", keywordWildcard, 200];
    rs = [[db dao] executeQuery:sql];
    while ([rs next])
    {
      Card* tmpCard = [[[Card alloc] init] autorelease];
      [tmpCard hydrate:rs];
      [cardList addObject: tmpCard];
    }
    [rs close];
  }
  
	return cardList;
}


//--------------------------------------------------------------------------
// NSString retrieveCsvCardIdsForTag
// Returns a CSV string of card Ids for a given tag
//--------------------------------------------------------------------------
+ (NSString*) retrieveCsvCardIdsForTag: (NSInteger)setId
{
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  NSString *sql = [[NSString alloc] initWithFormat:@"SELECT card_id FROM card_tag_link WHERE tag_id = '%d'",setId];
	FMResultSet *rs = [[db dao] executeQuery:sql];
	NSString *outputStr = [[[NSString alloc] init] autorelease];
  int cardId;
  BOOL firstTime = YES;
	while ([rs next])
  {
    cardId = [rs intForColumn:@"card_id"];
    if (firstTime == YES)
    {
      outputStr = [outputStr stringByAppendingFormat:@"%d",cardId];
      firstTime = NO;
    }
    else {
      outputStr = [outputStr stringByAppendingFormat:@",%d",cardId];
    }
  }
	[rs close];
  [sql release];
	return outputStr;
}


//--------------------------------------------------------------------------
// Card* retrieveCardWithSQL: sql
// Returns single card with SQL - assumes 1 record, if multiple, will take last record
//--------------------------------------------------------------------------
+ (Card*) retrieveCardWithSQL: (NSString*) sql
{
  int i = 0;
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  while ([db dao].inUse && i < 5)
  {
    LWE_LOG(@"Database is busy %d",i);
    usleep(100);
    i++;
  }
  FMResultSet *rs = [[db dao] executeQuery:sql];
  Card* tmpCard = [[[Card alloc] init] autorelease];
  while ([rs next])
  {
    [tmpCard hydrate:rs];
  }
  [rs close];
  return tmpCard;
}


//--------------------------------------------------------------------------
// Card* retrieveCardByPK: cardId
// Returns single card by PK
//--------------------------------------------------------------------------
+ (Card*) retrieveCardByPK: (NSInteger)cardId
{
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  NSString *sql = [[NSString alloc] initWithFormat:@""
      "SELECT c.card_id AS card_id,u.card_level as card_level,u.user_id as user_id,u.wrong_count as wrong_count,u.right_count as right_count,c.headword,c.headword_en,c.reading,c.romaji,ch.meaning "
      "FROM cards c INNER JOIN cards_html ch ON c.card_id = ch.card_id LEFT OUTER JOIN user_history u ON c.card_id = u.card_id AND u.user_id = '%d' "
      "WHERE c.card_id = ch.card_id AND c.card_id = '%d'",[settings integerForKey:@"user_id"], cardId];
  Card* tmpCard = [CardPeer retrieveCardWithSQL:sql];
	[sql release];
	return tmpCard;
}


//--------------------------------------------------------------------------
// Card* hydrateCardByPK: card
// Hydrates a card
//--------------------------------------------------------------------------
+ (Card*) hydrateCardByPK: (Card*) card
{
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  NSString *sql = [[NSString alloc] initWithFormat:@""
                   "SELECT c.card_id AS card_id,u.card_level as card_level,u.user_id as user_id,u.wrong_count as wrong_count,u.right_count as right_count,c.headword,c.headword_en,c.reading,c.romaji,ch.meaning "
                   "FROM cards c INNER JOIN cards_html ch ON c.card_id = ch.card_id LEFT OUTER JOIN user_history u ON c.card_id = u.card_id AND u.user_id = '%d' "
                   "WHERE c.card_id = ch.card_id AND c.card_id = '%d'",[settings integerForKey:@"user_id"], [card cardId]];
  FMResultSet *rs = [[db dao] executeQuery:sql];
  while ([rs next])
  {
    [card setHeadword:[rs stringForColumn:@"headword"]];
    [card setHeadword_en:[rs stringForColumn:@"headword_en"]];
    [card setReading:[rs stringForColumn:@"reading"]];
    [card setRomaji:[rs stringForColumn:@"romaji"]];
    [card setMeaning:[rs stringForColumn:@"meaning"]];
  }
  [rs close];
	[sql release];
	return card;
}


//--------------------------------------------------------------------------
// Card* retrieveCardByLevel: levelId setId: setId
// Returns single card from a set by a level
//--------------------------------------------------------------------------
+ (Card*) retrieveCardByLevel: (NSInteger)levelId setId: (NSInteger)setId withRandom: (NSInteger) randomNum
{
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  Card *tmpCard;
	NSString *sql, *sql2;
  sql = [[NSString alloc] initWithFormat:@""
      "SELECT card_level,u.card_id AS card_id,u.wrong_count as wrong_count,u.right_count as right_count "
      "FROM card_tag_link l, user_history u WHERE u.card_id = l.card_id AND l.tag_id = '%d' AND u.user_id = '%d' "
      "AND u.card_level = '%d' LIMIT 1 OFFSET %d",setId,[settings integerForKey:@"user_id"],levelId,randomNum];
  int cardId = 0;
  int cardLevel = 0;
  int wrongCount = 0;
  int rightCount = 0;
  FMResultSet *rs = [[db dao] executeQuery:sql];
  while ([rs next])
  {
    cardId     = [rs intForColumn:@"card_id"];
    cardLevel  = [rs intForColumn:@"card_level"];
    wrongCount = [rs intForColumn:@"wrong_count"];
    rightCount = [rs intForColumn:@"right_count"];
  }
  [rs close];
  sql2 =  [[NSString alloc] initWithFormat:@""
    "SELECT %d as card_level, %d AS wrong_count, %d AS right_count, %d AS user_id, * "
    "FROM cards WHERE card_id = '%d'",cardLevel,wrongCount,rightCount,[settings integerForKey:@"user_id"],cardId];
  FMResultSet *rs2 = [[db dao] executeQuery:sql2];
  tmpCard = [[[Card alloc] init] autorelease];
  while ([rs2 next])
  {
    [tmpCard hydrate:rs2];
  }
  [rs2 close];
	[sql release];
	return tmpCard;
}


//--------------------------------------------------------------------------
// NSMutableArray* retrieveCardSet: setId
// Returns an array of Cards for a given tag
//--------------------------------------------------------------------------
+ (NSMutableArray*) retrieveCardSetWithSQL: (NSString*) sql hydrate:(BOOL)hydrate
{
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  FMResultSet *rs = [[db dao] executeQuery:sql];
  Card* tmpCard;
  NSMutableArray *cardList = [[[NSMutableArray alloc] init] autorelease];
  int i = 0;
  while ([rs next])
  {
    i++;
    tmpCard = [[Card alloc] init];
    if (hydrate)
    {
      [tmpCard hydrate:rs];
    }
    else
    {
      [tmpCard setCardId:[rs intForColumn:@"card_id"]];
    }
    [cardList addObject: tmpCard];
    [tmpCard release];
  }
  [rs close];
  return cardList;
}


//--------------------------------------------------------------------------
// NSMutableArray* retrieveCardSetIds: setId
// Returns an array of Card Ids for a given tag
//--------------------------------------------------------------------------
+ (NSMutableArray*) retrieveCardSetIds: (NSInteger) tagId
{
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  LWEDatabase *db = [LWEDatabase sharedLWEDatabase];
  NSMutableArray *cardIdList = [[[NSMutableArray alloc] init] autorelease];
  for(int i = 0; i < 6; i++)
  {
    [cardIdList addObject:[[[NSMutableArray alloc] init] autorelease]];
  }
  NSString *sql = [[NSString alloc] initWithFormat:@"SELECT l.card_id AS card_id,u.card_level as card_level "
"FROM card_tag_link l LEFT OUTER JOIN user_history u ON u.card_id = l.card_id AND u.user_id = '%d' WHERE l.tag_id = '%d'",[settings integerForKey:@"user_id"],tagId];
  FMResultSet *rs = [[db dao] executeQuery:sql];
  while ([rs next])
  {
    [[cardIdList objectAtIndex:[rs intForColumn:@"card_level"]] addObject:[NSNumber numberWithInt:[rs intForColumn:@"card_id"]]];
  }
  [rs close];
  [sql release];
  return cardIdList;
}

//--------------------------------------------------------------------------
// NSMutableArray* retrieveCardIdsForTagId: tagId
// Returns an array of Cards ids for a given set
//--------------------------------------------------------------------------
+ (NSMutableArray*) retrieveCardIdsForTagId: (NSInteger)tagId
{
  NSString *sql = [[NSString alloc] initWithFormat:@"SELECT card_id FROM card_tag_link WHERE tag_id = '%d'",tagId];
  NSMutableArray *cardList = [CardPeer retrieveCardSetWithSQL:sql hydrate:NO];
  [sql release];
  return cardList;
}


//--------------------------------------------------------------------------
// NSMutableArray* retrieveCardSet: setId
// Returns an array of Cards for a given tag
//--------------------------------------------------------------------------
+ (NSMutableArray*) retrieveCardSet: (NSInteger) tagId
{
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  NSString *sql = [[NSString alloc] initWithFormat:@""
        "SELECT c.card_id AS card_id,u.user_id as user_id,u.card_level as card_level,u.wrong_count as wrong_count,u.right_count as right_count,c.headword,c.headword_en,c.reading,c.romaji,ch.meaning "
        "FROM cards c, cards_html ch, card_tag_link l, user_history u WHERE c.card_id = ch.card_id AND c.card_id = u.card_id AND "
        "c.card_id = l.card_id AND l.tag_id = '%d' AND u.user_id = '%d'",tagId,[settings integerForKey:@"user_id"]];
  NSMutableArray *cardList = [CardPeer retrieveCardSetWithSQL:sql hydrate:YES];
  [sql release];
	return cardList;
}


//--------------------------------------------------------------------------
// NSMutableArray* retrieveCardSetByLevel: setId levelId: levelId
// Returns an array of Cards for a given set and level
//--------------------------------------------------------------------------
+ (NSMutableArray*) retrieveCardSetByLevel: (NSInteger)setId levelId:(NSInteger)levelId
{
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  NSString *sql = [[NSString alloc] initWithFormat:@""
        "SELECT c.card_id AS card_id,u.user_id as user_id,u.card_level as card_level,u.wrong_count as wrong_count,u.right_count as right_count,c.headword,c.headword_en,c.reading,c.romaji,ch.meaning "
        "FROM cards c, cards_html ch, card_tag_link l, user_history u WHERE c.card_id = ch.card_id AND c.card_id = u.card_id "
        "AND c.card_id = l.card_id AND u.user_id = '%d' AND l.tag_id = '%d' AND u.card_level = '%d'",[settings integerForKey:@"user_id"],setId,levelId];
  NSMutableArray *cardList = [CardPeer retrieveCardSetWithSQL:sql hydrate:YES];
	[sql release];
	return cardList;
}


@end

File: android/src/com/longweekendmobile/android/xflash/model/CardPeer.java
package com.longweekendmobile.android.xflash.model;

//  CardPeer.java
//  Xflash
//
//  Created by Todd Presson on 1/17/12.
//  Copyright 2012 Long Weekend LLC.  All rights reserved.
//
//  public CardPeer() 
//
//      *** ALL METHODS STATIC ***
//
//  public Card blankCardWithId(int  )
//  public Card blankCard()
//  public boolean keywordIsReading(String  )
//  public boolean keywordIsHeadword(String  )
//  public ArrayList<Card> searchCardsForKeyword(String  )
//  public Card retrieveCardByPK(int  )
//
//  private String FTSSQLForKeyword(String  ,String  ,int  )
//  private Card retrieveCardWithCursor(Cursor  )
//  private ArrayList<Card> addCardsToList(ArrayList<Card>  ,Cursor  ,boolean  )
//
//  public ArrayList<ArrayList<Integer>> retrieveCardsSortedByLevelForTag(Tag  )
//  public ArrayList<Card> retrieveFaultedCardsForTag(Tag  )
//  public ArrayList<Card> retrieveCardSetForExampleSentenceId(int  )
//
//  public void setJ(boolean  )

import java.util.ArrayList;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.longweekendmobile.android.xflash.XFApplication;
import com.longweekendmobile.android.xflash.XflashSettings;
import android.util.Log;

public class CardPeer
{
    private static final String MYTAG = "XFlash CardPeer";
    
    // factory that cares about what language we are using
    public static Card blankCardWithId(int inId)
    {
        Card tempCard;

        if( XFApplication.IS_JFLASH )
        {
            tempCard = new JapaneseCard();
        }
        else
        {
            tempCard = new ChineseCard();
        }

        tempCard.setCardId(inId);
    
        return tempCard;    
    }


    // when we REALLY want a blank card
    public static Card blankCard()
    {
        Card tempCard = blankCardWithId(0);

        return tempCard;
    }


    // returns 'true' if the keyword appears to be reading-like.  note that
    // this implementation is specific to Chinese Flash, and would not work
    // for JFlash
    //
    // TODO - not tested in instances when it would return 'true'
    //        also, do we need a special unicode char variable?
    @SuppressWarnings("unused")
    public static boolean keywordIsReading(String inKey)
    {
        if( XFApplication.IS_JFLASH )
        {
            return false;
        }

        String[] keySplit = inKey.split(" ");

        // if any components are longer than 5 chars, it's not pinyin
        for( String key : keySplit )
        {
            if( key.length() > 5 )
            {
                return false;
            }
        }

        // okay, it has short bits, so make sure there's no kanji/hanzi
        boolean hasIdeograph = false;

        int tempInt = inKey.length();
        for(int charIndex = 0; charIndex < tempInt; charIndex++)
        {
            char testChar = inKey.charAt(charIndex);

            if(testChar >= 0x4E00)
            {
                hasIdeograph = true;
            }
        }

        return ( hasIdeograph == false );

    }  // end keywordIsReading()

    // TODO - see note on keywordIsReading() above
    @SuppressWarnings("unused")
    public static boolean keywordIsHeadword(String inKey)
    {
        if( XFApplication.IS_JFLASH )
        {
            return false;
        }

        // strip out numbers, symbols, and whitespace
        String keySplit[] = inKey.split("? ");

        for( String key : keySplit )
        {
            int tempInt = key.length();
            for(int charIndex = 0; charIndex < tempInt; charIndex++)
            {
                char testChar = key.charAt(charIndex);

                if( testChar < 0x2E80 || testChar > 0x9FFF )
                {
                    return false;
                }
            }   
        }

        return true;

    }  // end keywordIsHeadword()

    // returns an ArrayList of Card objects after searching for keyword
    // TODO - untested, need database with table cards_search_content
    public static ArrayList<Card> searchCardsForKeyword(String inKey)
    {
        SQLiteDatabase tempDB = XFApplication.getWritableDao();
        
        ArrayList<Card> cardList = new ArrayList<Card>();
        int queryLimit = 100;
        String column = null;

        // do not hydrate these cards, we will flywheel it on the table view
        if( keywordIsHeadword(inKey) )
        {
            column = "headword";
        }
        else if( keywordIsReading(inKey) )
        {
            column = "reading";
        }

        String query = FTSSQLForKeyword(inKey,column,queryLimit);
        
        Cursor myCursor = tempDB.rawQuery(query,null);
        myCursor.moveToFirst();

        cardList = addCardsToList(cardList,myCursor,false);
        myCursor.close();

        // now, did we get enough cards, query 'content' column as well
        // (if it was a no-column search, don't re-run
        if( column != null && cardList.size() < queryLimit )
        {
            int newLimit = cardList.size() - queryLimit;
            
            query = FTSSQLForKeyword(inKey,"content",newLimit);
            myCursor = tempDB.rawQuery(query,null);
                
            cardList = addCardsToList(cardList,myCursor,false);
        }

        return cardList;

    }  // end searchCardsForKeyword()

    
    public static Card retrieveCardByPK(int inId)
    {
        SQLiteDatabase tempDB = XFApplication.getWritableDao();
        
        int tempInt = XflashSettings.getCurrentUserId();

        String[] selectionArgs = new String[] { Integer.toString(tempInt), Integer.toString(inId) };
        String query = 
     "SELECT c.card_id AS card_id,u.card_level as card_level,u.user_id as user_id," +
     "u.wrong_count as wrong_count,u.right_count as right_count,c.*,ch.meaning " + 
     "FROM cards c INNER JOIN cards_html ch ON c.card_id = ch.card_id " + 
     "LEFT OUTER JOIN user_history u ON c.card_id = u.card_id AND u.user_id = ?" +
     "WHERE c.card_id = ch.card_id AND c.card_id = ?";

        Cursor myCursor = tempDB.rawQuery(query,selectionArgs);
        Card tempCard = retrieveCardWithCursor(myCursor);
        myCursor.close();

        return tempCard;        

    }  // end retrieveCardByPK()
   
 
    // this will return a SQL query that will return Card IDs from the FTS table
    // based on settings passed in
    // TODO - untested, we need database with table column card_search_content
    private static String FTSSQLForKeyword(String inKey,String inColumn,int inLimit)
    {
        // users understand '?' better than '*' ...  perhaps?
        String keywordWildcard = inKey.replace("*","?");
        String column;
        String searchString;

        if( inColumn != null )
        {
            // if we are searching for a specific column and not the
            // whole table, wrap it in quotes for exact match
            searchString = "\"" + keywordWildcard + "\"";
            column = inColumn;
        }
        else
        {
            searchString = keywordWildcard;
            column = "cards_search_content";
        }

        // do the search using SQLite FTS
        String query = "SELECT card_id FROM cards_search_content WHERE " + column + " MATCH '" + searchString + "' ORDER BY LENGTH(" + column + ") ASC, ptag DESC LIMIT " + inLimit;

        return query;

    }  // end FTSSQLForKeyword()

    // returns a single Card from the cursor, in case of multiple rows
    // will operate on the LAST one
    private static Card retrieveCardWithCursor(Cursor inCursor)
    {
        Card tempCard = blankCard();
    
        inCursor.moveToLast();
        tempCard.hydrate(inCursor);
        
        return tempCard;
    }


    // looping helper, populates an ArrayList of Cards, either lightweight or full
    private static ArrayList<Card> addCardsToList(ArrayList<Card> inList,Cursor inCursor,boolean shouldHydrate)
    {
        Card tempCard;
        int rowCount = inCursor.getCount();
        
        for(int i = 0; i < rowCount; i++)
        {
            if( shouldHydrate )
            {
                tempCard = blankCard();
                tempCard.hydrate(inCursor);
            }
            else
            {
                int tempColumn = inCursor.getColumnIndex("card_id");
                int tempInt = inCursor.getInt(tempColumn);

                tempCard = blankCardWithId(tempInt);
            }

            inList.add(tempCard);
            inCursor.moveToNext();
 
        } // end for loop

        return inList;

    }  // end addCardsToList()

    
    // returns an ArrayList of cards for the given Tag
    public static ArrayList<ArrayList<Card>> retrieveCardsSortedByLevelForTag(Tag inTag)
    {
        SQLiteDatabase tempDB = XFApplication.getWritableDao();
        
        // set up our empty array list to return
        ArrayList<ArrayList<Card>> cardList = new ArrayList<ArrayList<Card>>(6);
        for(int i = 0; i < 6; i++)
        {
            ArrayList<Card> tempList = new ArrayList<Card>();
            cardList.add(tempList);
        }
        
        String[] selectionArgs = new String[] { Integer.toString( XflashSettings.getCurrentUserId() ), Integer.toString( inTag.getId() ) };
        
        String query = "SELECT l.card_id AS card_id,u.card_level as card_level " +
                     "FROM card_tag_link l LEFT OUTER JOIN user_history u ON u.card_id = l.card_id " +
                     "AND u.user_id = ? WHERE l.tag_id = ?";
        
        Cursor myCursor = tempDB.rawQuery(query,selectionArgs);

        int rowCount = myCursor.getCount();
        myCursor.moveToFirst();

        for(int i = 0; i < rowCount; i++)
        {
            int tempColumn = myCursor.getColumnIndex("card_level");
            int tempLevelId = myCursor.getInt(tempColumn);

            tempColumn = myCursor.getColumnIndex("card_id");
            int tempCardId = myCursor.getInt(tempColumn);

            Card tempCard = blankCardWithId(tempCardId);
            tempCard.setLevelId(tempLevelId);

            cardList.get(tempLevelId).add(tempCard);

            myCursor.moveToNext();
        }

        myCursor.close();

        return cardList;  

    }  // end retrieveCardsSortedByLevelForTag()

    // returns an ArrayList of cardId integers contained in the Tag inTag
    public static ArrayList<Card> retrieveFaultedCardsForTag(Tag inTag)
    {
        SQLiteDatabase tempDB = XFApplication.getWritableDao();
        
        ArrayList<Card> tempList = new ArrayList<Card>();
        String[] selectionArgs = new String[] { Integer.toString( inTag.getId() ) };
        String query = "SELECT card_id FROM card_tag_link WHERE tag_id = ?";
    
        Cursor myCursor = tempDB.rawQuery(query,selectionArgs);
        myCursor.moveToFirst();
        addCardsToList(tempList,myCursor,false);
        myCursor.close();

        return tempList;

    }  // end retrieveFaultedCardsForTag()


    // returns an ArrayList<Integer> of cardIds that are linked to the sentence
    // 'inId' primary key of the sentence to look up cards for
    // TODO - uh... the code actually returns an array of Card objects, not IDs
    //      - as the original comment would imply
    public static ArrayList<Card> retrieveCardSetForExampleSentenceId(int inId)
    {
        SQLiteDatabase tempDB = XFApplication.getWritableDao();
        
        String query;
        String[] selectionArgs = new String[] { Integer.toString(inId) };
        ArrayList<Card> tempList = new ArrayList<Card>();

        // had to add 'meaning' to the query, else the program throws an
        // exception on Card.hydrate()
        query = "SELECT c.*, meaning FROM card_sentence_link l, cards c, cards_html WHERE l.card_id = c.card_id AND sentence_id = ? AND l.should_show = '1' AND cards_html.card_id = c.card_id";

        Cursor myCursor = tempDB.rawQuery(query,selectionArgs);
        int rowCount = myCursor.getCount();
        myCursor.moveToFirst();

        for(int i = 0; i < rowCount; i++)
        {
            Card tempCard = blankCard();
             
            tempCard.hydrate(myCursor,true);
            tempList.add(tempCard);

            myCursor.moveToNext();      
        }           

        myCursor.close();
        
        return tempList;

    }  // end retrieveCardSetForExampleSentenceId()


}  // end CardPeer declaration




